src/RecurringGrantDrop.sol

```diff
@@ -185,26 +199,96 @@ contract RecurringGrantDrop is Ownable2Step{
         );
     }
 
+    /// @notice Claim a reserved grant from the past
+    /// @param timestamp The timestamp of the reservation
+    /// @param receiver The address that will receive the tokens (this is also the signal of the ZKP)
+    /// @param root The root of the Merkle tree (signup-sequencer or world-id-contracts provides this)
+    /// @param nullifierHash The nullifier for this proof, preventing double signaling
+    /// @param proof The zero knowledge proof that demonstrates the claimer has a verified World ID
+    /// @dev hashToField function docs are in lib/world-id-contracts/src/libraries/ByteHasher.sol
+    function claimReserved(uint256 timestamp, address receiver, uint256 root, uint256 nullifierHash, uint256[8] calldata proof, bytes calldata signature)
+        external
+    {
+        uint256 grantId = grant.calculateId(timestamp);
+        checkClaimReserved(timestamp, receiver, root, nullifierHash, proof, signature);
+
+        nullifierHashes[nullifierHash] = true;
+
+        SafeERC20.safeTransferFrom(token, holder, receiver, grant.getAmount(grantId));
+
+        emit GrantClaimed(grantId, receiver);
+    }
+
+    /// @notice Check whether a claim is valid
+    /// @param timestamp The timestamp of the reservation
+    /// @param receiver The address that will receive the tokens (this is also the signal of the ZKP)
+    /// @param root The root of the Merkle tree (signup-sequencer or world-id-contracts provides this)
+    /// @param nullifierHash The nullifier for this proof, preventing double signaling
+    /// @param proof The zero knowledge proof, array of 8 uint256 elements, demonstrating that the claimer has a verified World ID
+    /// @param signature The off-chain signature of the reservation.
+    function checkClaimReserved(uint256 timestamp, address receiver, uint256 root, uint256 nullifierHash, uint256[8] calldata proof, bytes calldata signature)
+        public
+    {
+        uint256 grantId = grant.calculateId(timestamp);
+
+        if (receiver == address(0)) revert InvalidReceiver();
+        if (timestamp > block.timestamp) revert InvalidTimestamp();
+
+        this.checkNullifier(grantId, receiver, root, nullifierHash, proof);
+
+        grant.checkReservationValidity(timestamp);
+
+        address signer = ECDSA.recover(keccak256(abi.encode(timestamp, nullifierHash)), signature);
```

Comment:
Note the OpenZeppelin version has additional restrictions on the signature, which are not necessary for our application. It also doesn't hurt, as long as we take care of this at the signer side (otherwise ~50% of signatures will be rejected).